### PR TITLE
Conds can produce syntax errors

### LINKED ISSUE #172: "WHERE" and "AND" clauses can produce syntax errors
Hello!

First of all, thanks a lot for this amazing tool. I've been using it a lot in my personal projects.

Recently, I came across several issues when working with `WHERE` and `AND` clauses. Previously in #149 and `v1.30.1` you've added validation of arrays used in these queries. But they can still produce empty clauses if array is not empty but contains only empty strings.

Example:

```
sb.Distinct().Select("*").Where("")
sb.And("", "")
sb.Distinct().Select("*").Where(sb.And("", ""))
```

will produce accordingly:

```
SELECT DISTINCT * WHERE
( AND )
SELECT DISTINCT * WHERE ( AND )
```

In `v1.30.1` you've implemented counting size in bytes of strings to be inserted into `AND` condition. I think this might be handy in resolving this issue. If nothing can be added except for predefined constants, so i think nothing should be added. So the `And` function can be changed like this:

```
func (c *Cond) And(andExpr ...string) string {
	if len(andExpr) == 0 {
		return ""
	}

	exprByteLen := estimateStringsBytes(andExpr)
	if exprByteLen == 0 {
		return ""
	}

	buf := newStringBuilder()

	// Ensure that there is only 1 memory allocation.
	size := len(lparen) + len(rparen) + (len(andExpr)-1)*len(opAND) + exprByteLen
	buf.Grow(size)

	buf.WriteString(lparen)
	buf.WriteStrings(andExpr, opAND)
	buf.WriteString(rparen)
	return buf.String()
}
```

Same logic can be applied to `WHERE` clause, if all `andExprs` are empty, so don't generate them. So `AddWhereExpr` can be changed like this:

```
func (wc *WhereClause) AddWhereExpr(args *Args, andExpr ...string) *WhereClause {
	if len(andExpr) == 0 {
		return wc
	}

	andExprsBytesLen := 0
	for _, expr := range andExpr {
		andExprsBytesLen += len(expr)
	}

	if andExprsBytesLen == 0 {
		return wc
	}

	// Merge with last clause if possible.
	if len(wc.clauses) > 0 {
		lastClause := &wc.clauses[len(wc.clauses)-1]

		if lastClause.args == args {
			lastClause.andExprs = append(lastClause.andExprs, andExpr...)
			return wc
		}
	}

	wc.clauses = append(wc.clauses, clause{
		args:     args,
		andExprs: andExpr,
	})
	return wc
}
```

With these fixes examples provided previously will produce these values:
```
SELECT DISTINCT *
empty_string
SELECT DISTINCT *
```

If you think that these changes are legit and I'm not missing something, I can work on them in a PR. Examples above are just some raw ideas not the final code.

### PR DESCRIPTION
**Pull request by rodionovv**
fixes #172 

Everything discussed in #172 is added.

But i mentioned that solution discussed in #172 is not working for all cases. Checking the size of bytes to be written is working only when all input strings are empty. But if some of them are empty and some are not syntax errors will still be produced. So case like this:

`sb.And("", "1 = 1", "")`

will produce this sql string:

`( AND 1 = 1 AND )`

Since total length of input is not zero everything will be inserted. The best way to fix this issue i found is to edit `WriteStrings` function. It checks for the size of strings its writing to buffer and if it's empty it skips the step also ignoring separator.

Also i added validation of strings in all `Cond`s where string argument is present. All of them might produce syntax errors if string is empty. For example `EQ` will produce string like: ` = 1` if `field` is empty.

### CODE REVIEW

File: whereclause.go [hunk 1]
```diff
@@ -92,6 +92,12 @@ func (wc *WhereClause) AddWhereExpr(args *Args, andExpr ...string) *WhereClause
 		return wc
 	}
 
+	andExprsBytesLen := estimateStringsBytes(andExpr)
+
+	if andExprsBytesLen == 0 {
```
Comment (huandu) [inline whereclause.go:97]: Need some cases to cover this branch.

**Review by huandu — CHANGES_REQUESTED** [review]

**Comment by huandu** [comment]
Thanks for your contribution!

Your PR looks good to me in general. Just one comment: You can add some cases to test against the empty WHERE clause case, e.g. `Select("*").From("a").Where("", "")` should build `SELECT * FROM a`.

File: whereclause.go (hunk 1 above)
Comment (rodionovv) [reply whereclause.go:97]: I've added test covering those lines. But found another problem. 

Now building with `WHERE` clause can produce an extra blank space. In case `WHERE` was added through `AddWhereExpr` function. Since it can be only applied to an existing and non nil `WhereClause` it will be replaced with empty space during Build stage.

This problem can be  reproduced for both empty array and array of empty strings. I've added two test cases `TestEmptyStringsWhereAddWhereExpr` and `TestEmptyAddWhereExpr` to demonstrate how it can be achieved.

I felt that fixing this is not a part of current pr. I think its a bit another problem rather the one i was trying to fix here. I can still try to fix it in a current or maybe in a new pr. What do you think about it?

**Review by huandu — APPROVED** [review]
LGTM

**Comment by huandu** [comment]
Released in tag v1.31.0.
